### ISSUE TITLE
Model introspection for units

### ISSUE THREAD

**Issue by aieri** [canonical/operator#1436]
I see that the `ops.Model` class allows units to introspect the model they're in, but only in limited ways.
Allowing units to have a (read-only) full view of the model would allow charms to make more informed decisions about their behaviors. For example, applications that should be ubiquitous like grafana-agent would be able to tell if they have been deployed on some but not all machines and warn the user accordingly, without the need for external tooling.

In other words, the ops module could provide an idiomatic way for charms to consume `juju export-bundle` without awkward interactions with the juju api.

**Comment by tonyandrewmeyer** [comment]
Hi @aieri - the idea is that charms generally should only need to know about their specific concerns, rather than knowing about the overall state of the system. The leader often does have more information because it might need to do some coordination, but each individual unit should only need to know about its part of the world.

For the grafana-agent case, what would you (the charm) be doing if it wasn't installed somewhere? Without knowing the larger context, I would think that installation should be part of the initialisation (the `install` event for example) and that if it failed the unit would go into a non-active state (maybe error, maybe maintenance, depending on what was wrong). Why would other units need to know about that?

If there's a need to know that a unit has done something, typically you'd pass that information around via relation data.

Do you have a more specific example of where you'd want to use this?

**Comment by aieri** [comment]
hi @tonyandrewmeyer, yes, I am aware that traditionally charms aren't really able to know much beyond the boundaries of their relations. Would there really be a downside if they did, though?

For the grafana-agent case, the end goal is allowing an operator to validate if an environment is properly deployed (i.e. well designed) or not without needing an external tool. Currently, a charm can determine if it has been deployed on the wrong machine (i.e. ntp in an lxd container), but it cannot determine if it has _not_ been deployed where it should have been (where deployed = a human has decided that the unit should or should not be somewhere, not that installation failed). What should happen once that has been determined should depend on what the charm finds appropriate; in the grafana-agent case I think that a metric+alert would be useful and better than setting the status to blocked.

In general, I would like to be able to delegate model-level validation decisions to the charms themselves, but in this instance I cannot do so without providing the charms extra context.

I hope this is clearer.

**Comment by tonyandrewmeyer** [comment]
> I am aware that traditionally charms aren't really able to know much beyond the boundaries of their relations. Would there really be a downside if they did, though?

We want ops to make it easy to write great charms, fitting into the intended model. We don't want to add functionality that goes against the model.

> For the grafana-agent case, the end goal is allowing an operator to validate if an environment is properly deployed (i.e. well designed) or not without needing an external tool. Currently, a charm can determine if it has been deployed on the wrong machine (i.e. ntp in an lxd container), but it cannot determine if it has _not_ been deployed where it should have been (where deployed = a human has decided that the unit should or should not be somewhere, not that installation failed).

Does relation data not solve this for you? Each unit can put into a databag a "I have installed grafana-agent" flag and the leader can then do something with that in aggregate. Or if you're wanting the data in grafana, then can't the charm just send a 'grafana-agent-running' gauge or similar itself, and it's Grafana that takes care of any aggregation etc?

> In general, I would like to be able to delegate model-level validation decisions to the charms themselves, but in this instance I cannot do so without providing the charms extra context.

Charms aren't really meant to operate at the model level (app level for leaders, unit level for the others).

By the way, @benhoyt and I are also in the Hague this week - if it'd be easier, feel free to find some time with us in person to discuss this.

**Comment by benhoyt** [comment]
Thanks Tony. Agree with all of that -- this is going against the Juju model and very unlikely to be added. I'll close this for now, but we can always discuss solutions further (in person or otherwise).
